### PR TITLE
Primitive type

### PR DESCRIPTION
**Pull request by MingzhiYe16**
## Checklist
- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

File: metadata-ingestion/src/datahub/ingestion/source/thrift.py [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+        self, ctx: ThriftParser.Base_typeContext
+    ) -> SchemaFieldDataType:
+        return self.bind_SchemaFieldDataType_from_Real_base_type(ctx.real_base_type())
+
+    def bind_SchemaFieldDataType_from_Real_base_type(
+        self, ctx: ThriftParser.Real_base_typeContext
+    ) -> SchemaFieldDataType:
+        if ctx.TYPE_I32() or ctx.TYPE_I64() or ctx.TYPE_DOUBLE():
+            return SchemaFieldDataType(NumberType())
+        elif ctx.TYPE_STRING:
```
Comment (wizardxz) [inline metadata-ingestion/src/datahub/ingestion/source/thrift.py]: TYPE_STRING()

Please add unit tests which cover all primitive_types

**Review by wizardxz — CHANGES_REQUESTED** [review]

File: metadata-ingestion/build.gradle [hunk 2]
```diff
@@ -66,15 +72,23 @@ task lint(type: Exec, dependsOn: installDev) {
...
+    "isort src/ tests/ examples/ -s src/datahub/ingestion/source/dist/ && " +
+    "flake8 src/ tests/ examples/ --exclude src/datahub/ingestion/source/dist/ && " +
+    "mypy src/ tests/ examples/ --exclude src/datahub/ingestion/source/dist/"
+}
+
+task generateThriftGrammar(type: JavaExec) {
+  classpath = buildscript.configurations.classpath
+  mainClass = 'org.antlr.v4.Tool'
+  ignoreExitValue true
+  workingDir="src/datahub/ingestion/source"
```
Comment (rslanka) [inline metadata-ingestion/build.gradle]: can we use a temporary directory instead, that is deleted after ingestion is complete?

File: metadata-ingestion/build.gradle [hunk 3]
```diff
@@ -66,15 +72,23 @@ task lint(type: Exec, dependsOn: installDev) {
...
+    "source ${venv_name}/bin/activate && black --check --diff src/ tests/ examples/ --exclude src/datahub/ingestion/source/dist/ && isort --check --diff src/ tests/ examples/ -s src/datahub/ingestion/source/dist/ && flake8 --count --statistics src/ tests/ examples/ --exclude src/datahub/ingestion/source/dist/ && mypy src/ tests/ examples/ --exclude src/datahub/ingestion/source/dist/"
 }
 task lintFix(type: Exec, dependsOn: installDev) {
   commandLine 'bash', '-x', '-c',
     "source ${venv_name}/bin/activate && " +
-    "black src/ tests/ examples/ && " +
-    "isort src/ tests/ examples/ && " +
-    "flake8 src/ tests/ examples/ && " +
-    "mypy src/ tests/ examples/"
+    "black src/ tests/ examples/ --exclude src/datahub/ingestion/source/dist/ && " +
```
Comment (rslanka) [inline metadata-ingestion/build.gradle]: Can we eliminate this?

File: metadata-ingestion/examples/recipes/thrift_to_console.yml [hunk 4]
```diff
@@ -0,0 +1,10 @@
+---
+# see https://datahubproject.io/docs/metadata-ingestion/source_docs/file for complete documentation
+source:
+  type: "thrift"
+  config:
+    filename: "./examples/thrift_files/example.thrift"
```
Comment (rslanka) [inline metadata-ingestion/examples/recipes/thrift_to_console.yml:6]: Instead of a single file per recipe, can we accept a list of them?

File: metadata-ingestion/src/datahub/ingestion/source/thrift.py [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+    OtherSchema,
+    SchemaField,
+    SchemaFieldDataType,
+    SchemaMetadata,
+    StringType,
+)
+
+
+class ThriftSourceConfig(ConfigModel):
+    filename: str
```
Comment (rslanka) [inline metadata-ingestion/src/datahub/ingestion/source/thrift.py]: make this list instead?

File: metadata-ingestion/src/datahub/ingestion/source/thrift.py [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+    NumberType,
+    OtherSchema,
+    SchemaField,
+    SchemaFieldDataType,
+    SchemaMetadata,
+    StringType,
+)
+
+
+class ThriftSourceConfig(ConfigModel):
```
Comment (rslanka) [inline metadata-ingestion/src/datahub/ingestion/source/thrift.py]: Is there anything else that we want to make configurable?

File: docker/elasticsearch/env/docker.env [hunk 7]
```diff
@@ -1 +1 @@
-ES_JAVA_OPTS="-Xms256m -Xmx256m -Dlog4j2.formatMsgNoLookups=true"
+ES_JAVA_OPTS="-Xms256m -Xmx256m -Dlog4j2.formatMsgNoLookups=true"
```
Comment (rslanka) [inline docker/elasticsearch/env/docker.env]: Though looks good, doesn't seem related to this PR. Can we exclude this file?

File: metadata-ingestion/src/datahub/ingestion/source/Thrift.g4 [hunk 8]
```diff
@@ -0,0 +1,222 @@
+grammar Thrift;
```
Comment (rslanka) [inline metadata-ingestion/src/datahub/ingestion/source/Thrift.g4]: Please create a directory called `thrift` and move both thrift.py & Thrift.g4(lower-cased name) under that.

Comment (rslanka) [inline metadata-ingestion/src/datahub/ingestion/source/Thrift.g4]: Could you add a comment on this file is defining and if there are any current limitations, including a link to the documentation?

File: metadata-ingestion/src/datahub/ingestion/source/thrift.py [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+        for definition in ctx.definition():
+            yield from self.bind_MCPs_from_Definition(definition)
+
+    def bind_MCPs_from_Definition(
+        self, ctx: ThriftParser.DefinitionContext
+    ) -> Generator[MetadataChangeProposalWrapper, None, None]:
+        if ctx.struct_():
+            yield from self.bind_MCPs_from_struct_(ctx.struct_())
+        else:
+            raise NotImplementedError()
```
Comment (rslanka) [inline metadata-ingestion/src/datahub/ingestion/source/thrift.py]: Is this just defensive or are there incomplete unhandled cases? Please add a comment.

File: metadata-ingestion/build.gradle (hunk 2 above)
Comment (MingzhiYe16) [reply metadata-ingestion/build.gradle]: Sorry, I can't understand this, can you give me an example?

File: metadata-ingestion/src/datahub/ingestion/source/thrift.py [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+
+    def bind_MCPs_from_struct_(
+        self, ctx: ThriftParser.Struct_Context
+    ) -> Generator[MetadataChangeProposalWrapper, None, None]:
+
+        name = ctx.IDENTIFIER().getText()
+        urn = f"urn:li:dataset:(urn:li:dataPlatform:thrift,{name},PROD)"
```
Comment (rslanka) [inline metadata-ingestion/src/datahub/ingestion/source/thrift.py]: Please use https://github.com/datahub-project/datahub/blob/master/metadata-ingestion/src/datahub/emitter/mce_builder.py#L69 to construct the urn.

File: metadata-ingestion/src/datahub/ingestion/source/thrift.py [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+            nativeDataType=self.bind_nativeDataType_from_field_type(ctx.field_type()),
+        )
+
+    def bind_nativeDataType_from_field_type(
+        self, ctx: ThriftParser.Field_typeContext
+    ) -> str:
+        if ctx.base_type():
+            return ctx.getText()
+        else:
+            raise NotImplementedError()
```
Comment (rslanka) [inline metadata-ingestion/src/datahub/ingestion/source/thrift.py]: Comments about unhandled cases.

File: metadata-ingestion/src/datahub/ingestion/source/thrift.py [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+
+    def bind_SchemaFieldDataType_from_field_type(
+        self, ctx: ThriftParser.Field_typeContext
+    ) -> SchemaFieldDataType:
+        if ctx.base_type():
+            return self.bind_SchemaFieldDataType_from_base_type(ctx.base_type())
+        else:
+            raise NotImplementedError()
```
Comment (rslanka) [inline metadata-ingestion/src/datahub/ingestion/source/thrift.py]: comments about unhandled cases.

File: metadata-ingestion/src/datahub/ingestion/source/thrift.py [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+
+    def bind_SchemaFieldDataType_from_Real_base_type(
+        self, ctx: ThriftParser.Real_base_typeContext
+    ) -> SchemaFieldDataType:
+        if ctx.TYPE_I32() or ctx.TYPE_I64() or ctx.TYPE_DOUBLE():
+            return SchemaFieldDataType(NumberType())
+        elif ctx.TYPE_STRING:
+            return SchemaFieldDataType(StringType())
+        else:
+            raise NotImplementedError()
```
Comment (rslanka) [inline metadata-ingestion/src/datahub/ingestion/source/thrift.py]: comments about unhandled types.

File: metadata-ingestion/src/datahub/ingestion/source/thrift.py [hunk 14]
```diff
@@ -0,0 +1,141 @@
...
+
+
+@dataclass
+class ThriftSource(Source):
+    config: ThriftSourceConfig
+    report: SourceReport = field(default_factory=SourceReport)
```
Comment (rslanka) [inline metadata-ingestion/src/datahub/ingestion/source/thrift.py]: Are there any interesting metrics that we want to report, such as errors and processing time?

File: metadata-ingestion/src/datahub/ingestion/source/thrift.py [hunk 15]
```diff
@@ -0,0 +1,141 @@
...
+        config = ThriftSourceConfig.parse_obj(config_dict)
+        return cls(ctx, config)
+
+    def parse(
+        self, filename: str
+    ) -> Generator[MetadataChangeProposalWrapper, None, None]:
+        with open(filename) as text_file:
+            # lexer
+
+            lexer = ThriftLexer(InputStream(text_file.read()))
```
Comment (rslanka) [inline metadata-ingestion/src/datahub/ingestion/source/thrift.py]: Are there any exceptions in this block that we want to catch and report?

File: metadata-ingestion/src/datahub/ingestion/source/thrift.py [hunk 16]
```diff
@@ -0,0 +1,141 @@
...
+    def get_workunits(self) -> Iterable[Union[MetadataWorkUnit, UsageStatsWorkUnit]]:
+        for i, obj in enumerate(self.parse(self.config.filename)):
+            wu = MetadataWorkUnit(f"file://{self.config.filename}:{i}", mcp=obj)
+            self.report.report_workunit(wu)
+            yield wu
+
+    def get_report(self):
+        return self.report
+
+    def close(self):
```
Comment (rslanka) [inline metadata-ingestion/src/datahub/ingestion/source/thrift.py]: This is a good place to delete temporary directories.

File: metadata-ingestion/build.gradle (hunk 3 above)
Comment (MingzhiYe16) [reply metadata-ingestion/build.gradle]: No, because we don't want to lint the generated files

File: metadata-ingestion/examples/recipes/thrift_to_console.yml (hunk 4 above)
Comment (MingzhiYe16) [reply metadata-ingestion/examples/recipes/thrift_to_console.yml:6]: This requirement will be resolved in following diffs.

File: metadata-ingestion/src/datahub/ingestion/source/thrift.py (hunk 5 above)
Comment (MingzhiYe16) [reply metadata-ingestion/src/datahub/ingestion/source/thrift.py]: This requirement will be resolved in following diffs.

File: metadata-ingestion/src/datahub/ingestion/source/thrift.py (hunk 6 above)
Comment (MingzhiYe16) [reply metadata-ingestion/src/datahub/ingestion/source/thrift.py]: So far, no

File: docker/elasticsearch/env/docker.env (hunk 7 above)
Comment (MingzhiYe16) [reply docker/elasticsearch/env/docker.env]: Yes

File: metadata-ingestion/src/datahub/ingestion/source/Thrift.g4 (hunk 8 above)
Comment (MingzhiYe16) [reply metadata-ingestion/src/datahub/ingestion/source/Thrift.g4]: That's great

Comment (MingzhiYe16) [reply metadata-ingestion/src/datahub/ingestion/source/Thrift.g4]: OK

File: metadata-ingestion/src/datahub/ingestion/source/thrift.py (hunk 9 above)
Comment (MingzhiYe16) [reply metadata-ingestion/src/datahub/ingestion/source/thrift.py]: OK

File: metadata-ingestion/src/datahub/ingestion/source/thrift.py (hunk 10 above)
Comment (MingzhiYe16) [reply metadata-ingestion/src/datahub/ingestion/source/thrift.py]: OK

File: metadata-ingestion/src/datahub/ingestion/source/thrift.py (hunk 11 above)
Comment (MingzhiYe16) [reply metadata-ingestion/src/datahub/ingestion/source/thrift.py]: OK

File: metadata-ingestion/src/datahub/ingestion/source/thrift.py (hunk 12 above)
Comment (MingzhiYe16) [reply metadata-ingestion/src/datahub/ingestion/source/thrift.py]: OK

File: metadata-ingestion/src/datahub/ingestion/source/thrift.py (hunk 13 above)
Comment (MingzhiYe16) [reply metadata-ingestion/src/datahub/ingestion/source/thrift.py]: OK

File: metadata-ingestion/src/datahub/ingestion/source/thrift.py (hunk 14 above)
Comment (MingzhiYe16) [reply metadata-ingestion/src/datahub/ingestion/source/thrift.py:109]: This requirement will be resolved in following diffs.

File: metadata-ingestion/src/datahub/ingestion/source/thrift.py (hunk 15 above)
Comment (MingzhiYe16) [reply metadata-ingestion/src/datahub/ingestion/source/thrift.py:122]: This requirement will be resolved in following diffs.

File: metadata-ingestion/tests/unit/test_thrift.py [hunk 17]
```diff
@@ -0,0 +1,33 @@
...
+    pytestconfig,
+):
+
+    source = ThriftSource.create(
+        {"filename": "./examples/thrift_files/example.thrift"},
+        PipelineContext(run_id="test_run_id"),
+    )
+    mcp_objects = [wu.metadata.to_obj() for wu in source.get_workunits()]
+
+    with open(str(tmp_path / "thrift_mcps.json"), "w") as f:
```
Comment (rslanka) [inline metadata-ingestion/tests/unit/test_thrift.py]: Instead of the golden files based testing, please write tests against the in-memory python objects directly. See [this](https://github.com/datahub-project/datahub/blob/master/smoke-test/tests/test_stateful_ingestion.py) for an example. Feel free to ask for more if needed.

File: metadata-ingestion/build.gradle (hunk 3 above)
Comment (rslanka) [reply metadata-ingestion/build.gradle]: What I meant is that we generate the files outside in a temporary directory, not under src/datahub/ingestion/source/dist/.

File: metadata-ingestion/build.gradle (hunk 2 above)
Comment (MingzhiYe16) [reply metadata-ingestion/build.gradle]: we only need to generate once and it can be used for all input thrift files

Comment (rslanka) [reply metadata-ingestion/build.gradle]: You could use File.createTempDir and also specify that it be deleted on exit. Here is an example of how to do it.
https://github.com/sberbankjavaschool/homework/blob/11db332ee36bfa1f686842318a8ee68a7dd18c04/build.gradle#L77

File: metadata-ingestion/src/datahub/ingestion/source/thrift.py (hunk 14 above)
Comment (rslanka) [reply metadata-ingestion/src/datahub/ingestion/source/thrift.py]: You can create a sub-class of SourceReport, called ThriftReport to extend it.

File: metadata-ingestion/src/datahub/ingestion/source/thrift.py (hunk 16 above)
Comment (rslanka) [reply metadata-ingestion/src/datahub/ingestion/source/thrift.py]: This may no longer be needed if we specify tempDir.deleteOnExit in the gradle file.

File: metadata-ingestion/build.gradle (hunk 2 above)
Comment (rslanka) [reply metadata-ingestion/build.gradle]: If this task is building the lexer/parser etc thrift tools, we can even create the temporary directory under the project's buildDir.

**Review by rslanka — CHANGES_REQUESTED** [review]

File: metadata-ingestion/build.gradle (hunk 3 above)
Comment (MingzhiYe16) [reply metadata-ingestion/build.gradle]: OK, I'm working on this

File: metadata-ingestion/src/datahub/ingestion/source/thrift.py (hunk 14 above)
Comment (MingzhiYe16) [reply metadata-ingestion/src/datahub/ingestion/source/thrift.py]: Hello, I'm not very familiar with report, can you give me an example?

File: metadata-ingestion/tests/unit/test_thrift.py (hunk 17 above)
Comment (MingzhiYe16) [reply metadata-ingestion/tests/unit/test_thrift.py]: Hello, it sames that in the file you provided, the test is for checking whether pipline works correct, yes? In my test file, I don't use mysql pipline or any complex pipline, what should I check? I just read input file, ingest and compare with golden file.

Comment (MingzhiYe16) [reply metadata-ingestion/tests/unit/test_thrift.py]: Does checkoutpoint(in line 98 of [this](https://github.com/datahub-project/datahub/blob/8d281fc013080e1f955107774f4c0ce9639aab81/smoke-test/tests/test_stateful_ingestion.py#L98)) contains information about whether the pipline work well and we use this to check whether pipline work well?

Comment (rslanka) [reply metadata-ingestion/tests/unit/test_thrift.py]: In your test, you would be writing unit-tests directly against the key contents of the mcp_objects, such as the fieldPath, type and nativeDataType of the schemaFields produced. [This](https://github.com/datahub-project/datahub/blob/master/metadata-ingestion/tests/unit/test_kafka_source.py) is another example of such a test(we are generally trying to move away from golden file based tests, as they have a lot of irrelevant data that can change due to logic external to this connector). If you need help with a sample, I can help update this test itself, please let me know.

File: metadata-ingestion/src/datahub/ingestion/source/thrift.py (hunk 14 above)
Comment (rslanka) [reply metadata-ingestion/src/datahub/ingestion/source/thrift.py]: You can look at [this example](https://github.com/datahub-project/datahub/blob/master/metadata-ingestion/src/datahub/ingestion/source_report/sql/bigquery.py). There are more under [this](https://github.com/datahub-project/datahub/tree/master/metadata-ingestion/src/datahub/ingestion/source_report).

File: metadata-ingestion/tests/unit/test_thrift.py (hunk 17 above)
Comment (MingzhiYe16) [reply metadata-ingestion/tests/unit/test_thrift.py]: Should I still read the same example thrift file, but only check whether some critical features(check whether second field's nativetype is i32 and whether second field's type is {"com.linkedin.schema.NumberType": {}}) are right instead of compare with golden file?

File: metadata-ingestion/src/datahub/ingestion/source/thrift.py (hunk 14 above)
Comment (MingzhiYe16) [reply metadata-ingestion/src/datahub/ingestion/source/thrift.py]: I have created a sub-class ThriftReport and add variables: start time and end time. I found that failures{} has been in SourceReport, so I don't need to add errors{} to ThriftReport, yes?

File: metadata-ingestion/src/datahub/ingestion/source/thrift.py (hunk 15 above)
Comment (MingzhiYe16) [reply metadata-ingestion/src/datahub/ingestion/source/thrift.py]: Is it OK to do these here and add found errors to a metric of report: self.report.errors: List[str]?


                lexer = ThriftLexer(InputStream(text_file.read()))
                stream = CommonTokenStream(lexer)
                # parser
                parser = ThriftParser(stream)

                tree = parser.document()
                # evaluator
                yield from Binder().bind_MCPs_from_Document(tree)
        except Exception as e:
            self.report.errors.append(f"Error: {e}")`

File: metadata-ingestion/tests/unit/test_thrift.py (hunk 17 above)
Comment (rslanka) [reply metadata-ingestion/tests/unit/test_thrift.py]: Yes, that's correct. We basically don't want to use the golden files. It is perfectly okay to use as many thrift files as needed for the tests.

File: metadata-ingestion/src/datahub/ingestion/source/thrift.py (hunk 14 above)
Comment (rslanka) [reply metadata-ingestion/src/datahub/ingestion/source/thrift.py]: Correct. We could later extend it to get other fine-grained information like total number of files & structs that have been successfully processed, any metrics around ownership/documentation/tags/terms etc that we think are a useful summary to track.

File: docker/elasticsearch/env/docker.env (hunk 7 above)
Comment (wizardxz) [reply docker/elasticsearch/env/docker.env]: Can you address the comment?

File: metadata-ingestion/src/datahub/ingestion/source/Thrift.g4 (hunk 8 above)
Comment (wizardxz) [reply metadata-ingestion/src/datahub/ingestion/source/Thrift.g4]: Please address comments.

File: metadata-ingestion/src/datahub/ingestion/source/thrift.py [hunk 18]
```diff
@@ -0,0 +1,492 @@
...
+        return replace(self, namespaces={**self.namespaces, language: namespace})
+
+    def get(self, name: str) -> Optional[ResolveResult]:
+        return self.data.get(name)
+
+
+@dataclass(frozen=True)
+class GetHeader(ThriftVisitor):
+    filename: str
+    thrift_paths: Tuple[str, ...] = field(default_factory=tuple)
```
Comment (wizardxz) [inline metadata-ingestion/src/datahub/ingestion/source/thrift.py]: if thrift_paths is not used in GetHeader, please remove it

File: metadata-ingestion/src/datahub/ingestion/source/thrift.py (hunk 12 above)
Comment (wizardxz) [reply metadata-ingestion/src/datahub/ingestion/source/thrift.py]: Please address comments

**Review by wizardxz — CHANGES_REQUESTED** [review]

File: metadata-ingestion/src/datahub/ingestion/source/thrift.py (hunk 6 above)
Comment (MingzhiYe16) [reply metadata-ingestion/src/datahub/ingestion/source/thrift.py]: TODO

Comment (rslanka) [reply metadata-ingestion/src/datahub/ingestion/source/thrift.py]: TODO: add a directory option, potentially a git repo in the future.

**Comment by wizardxz** [comment]
Please merge it after adding TODO
